packages/block-library/src/gallery/tiles.native.js

```diff
@@ -0,0 +1,59 @@
+/**
+ * External dependencies
+ */
+import { View } from 'react-native';
+
+/**
+ * WordPress dependencies
+ */
+import { Children } from '@wordpress/element';
+
+const containerStyle = {
+	flexDirection: 'row',
+	flexWrap: 'wrap',
+};
+
+const tileStyle = {
+	overflow: 'hidden',
+	flexDirection: 'row',
+	alignItems: 'center',
+	borderColor: 'transparent',
+};
+
+function Tiles( props ) {
+	const {
+		columns,
+		children,
+		groutSpacing = 10,
+	} = props;
+
+	const tileCount = Children.count( children );
+	const lastTile = tileCount - 1;
+	const lastRow = Math.floor( lastTile / columns );
+
+	const wrappedChildren = Children.map( children, ( child, index ) => {
+		const row = Math.floor( index / columns );
+		const rowLength = row === lastRow ? ( lastTile % columns ) + 1 : columns;
+		const indexInRow = index % columns;
+
+		return (
+			<View style={ [ tileStyle, {
+				width: ( ( 1 / rowLength ) * 100 ) + '%',
+				borderLeftWidth: groutSpacing * ( indexInRow / rowLength ),
+				borderRightWidth: groutSpacing * ( 1 - ( ( indexInRow + 1 ) / rowLength ) ),
+				borderTopWidth: row === 0 ? 0 : groutSpacing / 2,
+				borderBottomWidth: row === lastRow ? 0 : groutSpacing / 2,
```

Comment:
Why are we setting these as borders and not margins?

Reply:
I used borders for this so that fixed sizes would be included in the total width of each item, preserving the correct wrapping behavior. I've added a note about this in the code comment.

Reply:
I didn't realize I had said margins here when I finished the review. Why not paddings? I would think they offer the same effect but they would be more meaningful than transparent borders

Reply:
I think padding is semantically better than border for what we are doing here, and I even used padding in an earlier iteration (from what I can tell it works pretty much the same way). The reason I switched back to using border is that I had the thought that border would be less likely to be used later if we wanted to modify the styles, so it was more "disposable" in that sense (i.e. by not using padding, we still have the opportunity to use it for its original purpose, if that need arises).

Reply:
> (i.e. by not using padding, we still have the opportunity to use it for its original purpose, if that need arises).

Could you help me understand this better? What kind of need for example? Using borders for spacing purposes is not very intuitive for other people who will maintain it. And having the need to put such a big code comment can be signaling that we are introducing a "hard to understand" code. So I'd prefer using paddings unless we have good reasons.

Reply:
This looks simpler on web, only right and bottom margins are used to give the spacing and the bottom margin is a constant. Only thing that is varies is the right margin, it is just set to 0 for the right most tile.

![galleryTiles2](https://user-images.githubusercontent.com/5032900/68482323-5da2b080-024a-11ea-8f23-53d2777bea7e.gif)

Can we use a similar approach?

Reply:
Hi @pinarol :wave: thanks for taking a look at this PR.

> Could you help me understand this better? What kind of need for example?

We are currently using borders to evenly distribute space between tiles. If we want (in the future) to give each tile padding, we can, since that property is not being used for this layout purpose.

> And having the need to put such a big code comment can be signaling that we are introducing a "hard to understand" code. So I'd prefer using paddings or margins unless we have good reasons.

FWIW, the use of `padding` instead of `border` won't alleviate the need for the comment. AFAIK, the use of margins is not possible for this. The aim of the comment is to make it _easier_ to understand for future maintainers.

> This looks simpler on web, only right and bottom margins are used to give the spacing and the bottom margin is a constant. Only thing that varies is the right margin, it is just set to 0 for the right most tile.

Right, web is able to use margins. We are not able to do this, since margins won't be included in the width calculations (breaking the wrapping behavior).

Reply:
> Right, web is able to use margins. We are not able to do this, since margins won't be included in the width calculations (breaking the wrapping behavior).

I see but It is possible if we change the approach a bit, it just requires us to calculate the tile width in pixels rather than percentages. 

Actually I think the most challenging part of this manual calculation is finding the `rowLength`, and you already did it for "calculating the width based on percentage" solution.

I still think we can make the current code easier to understand and maintain.

I'll drop my example here too, for documenting purposes. Let's give this a try and see how it goes. I added a round down in 0,00 precision so I think that'll help with exceeding the container width.

```

class Tiles extends Component {

	constructor() {
		super( ...arguments );
		this.onLayout = this.onLayout.bind( this );

		this.state = {
			containerWidth: 0,
		};
	}
		
	onLayout( event ) {
		const { width, height } = event.nativeEvent.layout;
		this.setState( {
			containerWidth: width,
		} );
	}

	render() {
		const {
			columns,
			children,
			spacing = 10,
		} = this.props;
		
		const tileCount = Children.count( children );
		const lastTile = tileCount - 1;
		const lastRow = Math.floor( lastTile / columns );

		const wrappedChildren = Children.map( children, ( child, index ) => {
			const { containerWidth } = this.state;
			if ( containerWidth === 0 ) {
				return null;
			}

			const rowLength = (() => {
				const row = Math.floor( index / columns );
				const rowLength = row === lastRow ? ( lastTile % columns ) + 1 : columns;	
				return rowLength;
			})();

			const tileWidth = (() => {
				const totalHorizontalMargin = ( rowLength - 1 ) * spacing;
				const width =  ( containerWidth - totalHorizontalMargin ) / rowLength;
				return Math.floor( width * 100 ) / 100; // round down to 2 decimals
			})();

			const isRightMostTile = ( index === lastTile ) || (( index + 1 ) % rowLength === 0);

			return (
				<View style={ [ styles.tileStyle, {
					width: tileWidth,
					marginRight: isRightMostTile ? 0 : spacing,
					marginBottom: spacing,
				} ] }>
					{ child }
				</View>
			);
		} );
	
		return (
			<View style={ styles.containerStyle } onLayout={ this.onLayout }>
				{ wrappedChildren }
			</View>
		);
	}
}

```

Reply:
> Let's give this a try and see how it goes.

I tried this out, but I'm still seeing problems on orientation changes, and columns (in landscape).

**Screenshots:**

Rotation:
<img src="https://user-images.githubusercontent.com/8507675/68641371-53d4c200-0556-11ea-92af-d6627e1ffc05.gif" width="360">

Columns in landscape:
<img src="https://user-images.githubusercontent.com/8507675/68641374-58997600-0556-11ea-8149-8d03ce584f92.gif" width="360">

Reply:
It looks like I made a mistake on this line:

`const isRightMostTile = ( index === lastTile ) || (( index + 1 ) % rowLength === 0);`

The correct one should be:

`const isRightMostTile = ( index === lastTile ) || (( index + 1 ) % columns === 0);`

It is working fine with this fix according to my experiments.

BTW I am fine with both:
- going with percentages and using paddings
 or 
- going with margins and calculating the width in pixels

I just wouldn't prefer to go with transparent borders as it is not what they are for and for that reason it is making code harder to understand. And we can always iterate if a special need comes up that will require us to use borders for spacing.

When it comes to paddings vs margins, they have both pros/cons. Advantage of padding solution is we don't need the onLayout handling. 

Advantage of margin solution is, it is making the space calculation very easy, the only thing to calculate is left as: `marginRight: isRightMostTile ? 0 : spacing,`. 

Personally I'd go with margins but I am fine with both solutions. 👍 

I wouldn't expect any performance differences between both ways because whenever container width changes the tiles should be re-rendered whether they are calculated in pixels or percentages.

Reply:
Hi @pinarol :wave: 

> The correct one should be:
> 
> `const isRightMostTile = ( index === lastTile ) || (( index + 1 ) % columns === 0);`
> 

I've tried the margin approach again with these suggested changes, however, I'm still seeing the same layout problem on orientation changes [described previously](https://github.com/WordPress/gutenberg/pull/18130#discussion_r345010804).

> BTW I am fine with both:
> 
> * going with percentages and using paddings
>   or
> * going with margins and calculating the width in pixels
> 
> I just wouldn't prefer to go with transparent borders as it is not what they are for and for that reason it is making code harder to understand. And we can always iterate if a special need comes up that will require us to use borders for spacing.
> 

Unfortunately, neither border nor padding are semantically correct in how they're being used here, according to the box model. I don't have a strong opinion about which is "better" between `border` and `padding`, so I've changed it to padding.

> When it comes to paddings vs margins, they have both pros/cons. Advantage of padding solution is we don't need the onLayout handling.
> 
> Advantage of margin solution is, it is making the space calculation very easy, the only thing to calculate is left as: `marginRight: isRightMostTile ? 0 : spacing,`.
> 
> Personally I'd go with margins but I am fine with both solutions. 
> 
> I wouldn't expect any performance differences between both ways because whenever container width changes the tiles should be re-rendered whether they are calculated in pixels or percentages.

Another advantage to the current spacing interpolation approach is that it allows `Tiles` to remain a stateless functional component instead of a stateful class component.

Since the `onLayout` approach seems to result in an unstable layout during orientation changes, I've gone with the interpolation approach. I've pushed a change to use `padding` instead of `border`, and updated the comment to reflect that.

Reply:
> I've tried the margin approach again with these suggested changes, however, I'm still seeing the same layout problem on orientation changes described previously.

I am super curious why that's happening now because I was able to repro the problem and that change seemed fixing that. If you want we can work on the problem together. But as I said before I am also fine with using paddings. 

> Unfortunately, neither border nor padding are semantically correct in how they're being used here

You have a point, but I wouldn't consider paddings and borders as _equally_ incorrect if they are used for spacing purposes. And if we are looking for semantically the most correct one that seems to be margins.

> Since the onLayout approach seems to result in an unstable layout during orientation changes, I've gone with the interpolation approach. 

If you mean [this](https://github.com/WordPress/gutenberg/pull/18130#discussion_r345010804) problem by unstable layout I think that's not because of onLayout, that one is just a result of an immature solution(I spent only ~30mins) and the logic that calculates the right most tile can have flaws, it just needs some tackling. Or is there any other problem that I am missing?

Reply:
> Or is there any other problem that I am missing?

I'm not sure. fwiw, your logic to determine the right-most tile looks correct to me. In fact, I had earlier implementations that used a similar approach with `margin`, and even `flex-basis`, but all of them suffered from layout problems on orientation changes. The only solution that seems robust in my tests is the current interpolation approach. When using the `onLayout` approach, adding `console.log(width)` showed some differences in the layout calculations:

```
Starting app in landscape:

2x 358.5454406738281 tiles.native.js:35
5x 360.7272644042969 tiles.native.js:35

After orientation change:

3x 548               tiles.native.js:35
3x 546.1818237304688 tiles.native.js:35
```

For each, `onLayout` is called 7 and 6 times, respectively, for portrait and landscape orientations. It seems that the width "adjusts" as the calculation "settles" on a final result. In the case of portrait, width seems to grow slightly, whereas in landscape it seems to shrink.

Reply:
I am all fine with going with padding solution 👍 onLayout can be called multiple times but we don't need to update the component state if container width stays same. And if container width does change, the tiles will need layout again independent from whether they are calculated using pixels or percentages. 

But if you observed some odd behavior in case of margins let's go with paddings. 

I am up for both: going as is or working together to see what's wrong with margin solution(because I couldn't repro with my local env).